net: tcp: Fix context leak during connect

neja20 opened the pull request:
Replaced tcp_out() with tcp_out_ext() when sending SYN during tcp connect, so we can intercept error value, beacuse in such case the packet would not be sent at all and the stack would not trigger any mechanism to flush this context so it ends up leaking. These scenarios can arise when the underlaying interface is not properly configured or is disconnected. The conn is marked to be closed and is closed before exiting tcp_in().
Since we can now idefntify this case we can also exit early in the net_tcp_connect() function and not wait for any timeout.

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -2922,11 +2922,16 @@ static enum net_verdict tcp_in(struct tcp *conn, struct net_pkt *pkt)
 			verdict = NET_OK;
```
Typo in commit msg `s/idefntify/identify/`

jukkar reviewed (commented):
LGTM, small nit about typo in commit msg

rlubos commented on subsys/net/ip/tcp.c:
```diff
@@ -3918,7 +3923,10 @@ int net_tcp_connect(struct net_context *context,
+			ret = -errno;
+			goto out;
+		} else if ((K_TIMEOUT_EQ(timeout, K_NO_WAIT)) &&
 		    conn->state != TCP_ESTABLISHED) {
```
nit: Please fix the alignment here

neja20 replied:
I hope I understood you correctly, you wanted the if statement aligned.

rlubos reviewed (commented):
Wow, that's a really good catch! Glad to see we have community on watch here, one more corner-case in our TCP stack covered.

Just please take a look at the nits reported and we're good.

dnejezchleb committed:
net: tcp: Fix context leak during connect

Replaced tcp_out() with tcp_out_ext() when sending
SYN during tcp connect, so we can intercept error value,
beacuse in such case the packet would not be sent at all
and the stack would not trigger any mechanism to flush
this context so it ends up leaking. These scenarios can
arise when the underlaying interface is not properly
configured or is disconnected. The conn is marked to be
closed and is closed before exiting tcp_in().
Since we can now identify this case we can also exit
early in the net_tcp_connect() function and not wait
for any timeout.

Signed-off-by: Daniel Nejezchleb <dnejezchleb@hwg.cz>
